Connect to api

elasfarc opened the pull request:
In this PR, I did the following:

Persist Add and Remove actions to the server 
Add loading feature when fetching data from the server

elasfarc committed:
Add API middleware

elasfarc committed:
Add API action creators

elasfarc committed:
Add action logic for loading books from the server

elasfarc committed:
Remove a book functionality

by dispatching two action:
- delete the specific book from the server.
- delete the book locally from the store without loading all the updated books from the server.

elasfarc committed:
Add a new book to the server

elasfarc committed:
Refactor Book reducer

since generating a new book unique id is no longer handled by the reducer --> there's no need to keep track of the 'lastID' variable through closure

elasfarc committed:
Book store keeps tracking of book's category instead of author

elasfarc committed:
Add API config file

elasfarc committed:
Add loading feature when fetching data from the server

elasfarc committed:
refactor Library && Fix linters errors

Emmyn5600 commented on src/components/AddNewBook.js:
```diff
@@ -20,12 +20,12 @@ const NewBook = () => {
-    dispatch(actions.addBook({ title: bookTitle, author: bookAuthor }));
+    dispatch(actions.addBook({ title: bookTitle, category: bookCategory }));
     resetInputs();
   };
```
- [ ] Nice you have implemented connect API successfully, however could you please validate title and category inputs so that it doesn't allow the user to add empty data.
![api](https://user-images.githubusercontent.com/55635977/130919512-2bcf6e3e-56c5-4a26-9206-d9209cefa4f0.png)

Emmyn5600 commented on src/redux/middleware/api.js:
```diff
@@ -0,0 +1,33 @@
+// eslint-disable-next-line consistent-return
+const api = ({ dispatch }) => (next) => async (action) => {
+  if (action.type !== apiActions.API_CALL_REQUESTED) return next(action);
+
```
- [ ] Could you please correct the eslint errors without disabling them.

![api1](https://user-images.githubusercontent.com/55635977/130919938-13b1846b-ed19-4997-84ee-611d7b836b6e.png)

- [ ]  Please use the `componentDidMount` lifecycle method or with react hooks requests. see the screenshot :point_down:

![api2](https://user-images.githubusercontent.com/55635977/130919908-7e393325-e59c-4396-a6ec-64cb2b2b8263.png)

elasfarc replied:
I didn't get the last point, could you please explain it

elasfarc replied:
![07663F3B-6546-4E4D-906F-23D4ADA8445F](https://user-images.githubusercontent.com/63879921/130928266-699744c8-6b54-4060-b3c4-3cffe9bf0754.jpeg)

Emmyn5600 reviewed (changes requested):
## STATUS: CHANGES REQUESTED :recycle:

Hi, @elasfarc 

Great work on implementing the project requirements, however there are some issues to be solved before you get approved but you are almost there:muscle:. 
   
   - The repo looks professional.:heavy_check_mark:
- Git/GitHub flow best practices have been followed:heavy_check_mark:
- There are no linter errors:heavy_check_mark:
- You have descriptive PR:heavy_check_mark:

Changes required :writing_hand: 
 
      * #### Please check the comments under the review.
       
Cheers and Happy coding! :clap: :clap: :clap:

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

#### _**Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise.**_
------------------------------------------------------------------------------------------------------------------------------
As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using this [form.](https://airtable.com/shrQAqnBwek5a0O0s)

elasfarc committed:
ADD validation: [title] and [category] inputs

so that it doesn't allow the user to add empty  data.

elasfarc committed:
Fix ESlint error

bellom reviewed (approved):
## Hi @elasfarc  👏  

**Great work making all the necessary corrections from the previous review.**

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 

### Congratulations! 🎉 

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

elasfarc committed:
[Libray] [Input-errors]

Add a general-purpose input error checker && fix when the page loads for the first time the submit button is not disabled

elasfarc committed:
Update AddNewBook.js
